doc/man/pam_pwquality.8.pod
=pod

=head1 NAME

pam_pwquality - PAM module to perform password quality checking

=head1 SYNOPSIS

B<pam_pwquality.so> [I<...>]

=head1 DESCRIPTION

This module can be plugged into the B<password> stack of a given service
to provide some plug-in strength-checking for passwords.
The code was originally based on pam_cracklib module and the module is
backwards compatible with its options.

The action of this module is to prompt the user for a password and check
its strength against a system dictionary and a set of rules for identifying
poor choices.

The first action is to prompt for a single password, check its strength
and then, if it is considered strong, prompt for the password a second time
(to verify that it was typed correctly on the first occasion). All being
well, the password is passed on to subsequent modules to be installed as the
new authentication token.

The checks for strength are:

=over 4

=item Palindrome

Is the new password a palindrome?

=item Case Change Only

Is the new password the same as the old one with only a change of case?

=item Similar

Is the new password too much like the old one? This is primarily controlled
by one argument, B<difok> which is a number of character changes (inserts,
removals, or replacements) between the old and new password that are enough
to accept the new password.

=item Simple

Is the new password too small? This is controlled by 6 arguments
B<minlen>, B<maxclassrepeat>, B<dcredit>, B<ucredit>, B<lcredit>,
and B<ocredit>. See the section on the arguments for the details of how
these work and there defaults.

=item Rotated

Is the new password a rotated version of the old password?

=item Same consecutive characters

Optional check for same consecutive characters.

=item Too long monotonic character sequence

Optional check for too long monotonic character sequence.

=item Contains user name

Check whether the password contains the user's name in some form.

=item Dictionary check

The I<Cracklib> routine is called to check if the password is part of
a dictionary.

=back

These checks are configurable either by use of the module arguments
or by modifying the F</etc/security/pwquality.conf> configuration file. The
module arguments override the settings in the configuration file.

=head1 OPTIONS

=over 4

=item B<debug>

This option makes the module write information to L<syslog(3)>
indicating the behavior of the module (this option does not write password
information to the log file).

=item B<authtok_type=>I<XXX>

The default action is for the module to use the following prompts when
requesting passwords: C<"New UNIX password: "> and
C<"Retype UNIX password: ">. The example word
I<UNIX> can be replaced with this option, by default it is empty.

=item B<retry=>I<N>

Prompt user at most I<N> times before returning with error. The default is
I<1>.

=item B<difok=>I<N>

This argument will change the default of I<1> for the number of changes in
the new password from the old password.

The special value of I<0> disables all checks of similarity of the new password
with the old password except the new password being exactly the same as
the old one.

=item B<minlen=>I<N>

The minimum acceptable size for the new password. In addition to the number of
characters in the new password, credit (up to the individual credit setting in
length) is given for each different kind of character (I<other>, I<upper>,
I<lower> and I<digit>). Set the individual I<credit> settings to 0 or negative
to have I<minlen> count as true length of the new password in characters.

The default for this parameter is I<8>.

Note that there is a pair of length limits also in I<Cracklib>, which is used
for dictionary checking, a "way too short" limit of I<4> which is hard coded
in and a build time defined limit (I<6>) that will be checked without
reference to B<minlen>.

=item B<dcredit=>I<N>

(N >= 0) This is the maximum credit for having digits in the new password.
If you have less than or I<N> digits, each digit will count +1 towards meeting
the current B<minlen> value. The default for B<dcredit> is I<0>
which means there is no bonus for digits in password.

(N < 0) This is the minimum number of digits that must be met for a new
password.

=item B<ucredit=>I<N>

(N >= 0) This is the maximum credit for having upper case letters in the new password.
If you have less than or I<N> upper case letters, each upper case letter will count +1 towards meeting
the current B<minlen> value. The default for B<ucredit> is I<0>
which means there is no bonus for upper case letters in password.

(N < 0) This is the minimum number of upper case letters that must be met for a new
password.

=item B<lcredit=>I<N>

(N >= 0) This is the maximum credit for having lower case letters in the new password.
If you have less than or I<N> lower case letters, each lower case letter will count +1 towards meeting
the current B<minlen> value. The default for B<lcredit> is I<0>
which means there is no bonus for lower case letters in password.

(N < 0) This is the minimum number of lower case letters that must be met for a new
password.

=item B<ocredit=>I<N>

(N >= 0) This is the maximum credit for having other characters in the new password.
If you have less than or I<N> other characters, each other character will count +1 towards meeting
the current B<minlen> value. The default for B<ocredit> is I<0>
which means there is no bonus for other characters in password.

(N < 0) This is the minimum number of other characters that must be met for a new
password.

=item B<minclass=>I<N>

The minimum number of required classes of characters for the new password.
The four classes are digits, upper and lower letters and other characters.
The difference to the B<credit> check is that a specific class of
characters is not required. Instead I<N> out of four of the classes are
required. By default the check is disabled.

=item B<maxrepeat=>I<N>

Reject passwords which contain more than I<N> same consecutive characters.
The default is 0 which means that this check is disabled.

=item B<maxsequence=>I<N>

Reject passwords which contain monotonic character sequences longer than I<N>.
The default is 0 which means that this check is disabled.
Examples of such sequence are '12345' or 'fedcb'. Note that
most such passwords will not pass the simplicity check unless the sequence
is only a minor part of the password.

=item B<maxclassrepeat=>I<N>

Reject passwords which contain more than I<N> consecutive characters of the
same class. The default is 0 which means that this check is disabled.

=item B<gecoscheck=>I<N>

If nonzero, check whether the individual words longer than 3 characters
from the L<passwd(5)> GECOS field of the user are contained in the new
password. The default is 0 which means that this check is disabled.

=item B<dictcheck=>I<N>

If nonzero, check whether the password (with possible modifications)
matches a word in a dictionary. Currently the dictionary check is performed
using the I<cracklib> library. The default is 1 which means that this check
is enabled.

=item B<usercheck=>I<N>

If nonzero, check whether the password (with possible modifications)
contains the user name in some form. The default is 1 which means that
this check is enabled. It is not performed for user names shorter
than 4 characters.

=item B<usersubstr=>I<N>

If greater than 3 (due to the minimum length in usercheck), check whether the
password contains a substring of the user name of at least I<N> length in some form.
The default is 0, which means this check is disabled.

=item B<enforcing=>I<N>

If nonzero, reject the password if it fails the checks, otherwise
only print the warning. The default is 1 which means that the weak password
is rejected (for non-root users).

=item B<badwords=>I<< <list of words> >>

The words more than 3 characters long from this space separated list are
individually searched for and forbidden in the new password.
By default the list is empty which means that this check is disabled.

=item B<dictpath=>I</path/to/dict>

This options allows for specification of non-default path to the cracklib
dictionaries.

=item B<enforce_for_root>

The module will return error on failed check even if the user changing the
password is root. This option is off by default which means that just
the message about the failed check is printed but root can change
the password anyway. Note that root is not asked for an old password
so the checks that compare the old and new password are not performed.

=item B<local_users_only>

The module will not test the password quality for users that are not present
in the F</etc/passwd> file. The module still asks for the password so
the following modules in the stack can use the B<use_authtok> option.
This option is off by default.

=item B<use_authtok>

This argument is used to I<force> the module to not prompt the user for
a new password but use the one provided by the previously stacked
B<password> module.

=item B<allowclasses=>I<< "ludo" >>

The list of letters that represent allowed character classes that
are allowed in a password (digits, uppercase, lowercase, others).
"d" is for digits, "u" is for uppercase letters, "l" is for
lowercase letters, "o" is for other characters.

=back

=head1 MODULE TYPES PROVIDED

Only the B<password> module type is provided.

=head1 RETURN VALUES

=over 4

=item PAM_SUCCESS

The new password passes all checks.

=item PAM_AUTHTOK_ERR

No new password was entered, the username could not be determined or the
new password fails the strength checks.

=item PAM_AUTHTOK_RECOVERY_ERR

The old password was not supplied by a previous stacked module or got not
requested from the user. The first error can happen if B<use_authtok>
is specified.

=item PAM_SERVICE_ERR

A internal error occurred.

=back

=head1 EXAMPLES

For an example of the use of this module, we show how it may be stacked
with the password component of L<pam_unix(8)>.

 #
 # These lines stack two password type modules. In this example the
 # user is given 3 opportunities to enter a strong password. The
 # "use_authtok" argument ensures that the pam_unix module does not
 # prompt for a password, but instead uses the one provided by
 # pam_pwquality.
 #
 password required pam_pwquality.so retry=3
 password required pam_unix.so use_authtok

Another example is for the case that you want to use sha256 password
encryption:

 #
 # These lines allow modern systems to support passwords of at least 14
 # bytes with extra credit of 2 for digits and 2 for others the new
 # password must have at least three bytes that are not present in the
 # old password
 #
 password required pam_pwquality.so \
               difok=3 minlen=15 dcredit=2 ocredit=2
 password required pam_unix.so use_authtok nullok sha256

And here is another example in case you don't want to use credits:

 #
 # These lines require the user to select a password with a minimum
 # length of 8 and with at least 1 digit number, 1 upper case letter,
 # and 1 other character
 #
 password required pam_pwquality.so \
               dcredit=-1 ucredit=-1 ocredit=-1 lcredit=0 minlen=8
 password required pam_unix.so use_authtok nullok sha256

=head1 SEE ALSO

L<pwscore(1)>, L<pwquality.conf(5)>, L<pam_pwquality(8)>,
L<pam.conf(5)>, L<PAM(8)>

=head1 AUTHORS

Tomas Mraz <tm@t8m.info>

Original author of B<pam_cracklib> module Cristian Gafton <gafton@redhat.com>


src/pam_pwquality.c
/*
 * PAM module for password quality checking using libpwquality
 *
 * See the end of the file for Copyright and License Information
 */

#include "config.h"

#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <ctype.h>
#include <limits.h>
#include <syslog.h>
#include <stdio.h>
#include <pwd.h>
#include <errno.h>
#include "pwquality.h"

/*
 * here, we make a definition for the externally accessible function
 * in this file (this definition is required for static a module
 * but strongly encouraged generally) it is used to instruct the
 * modules include file to define the function prototypes.
 */

#define PAM_SM_PASSWORD

#include <security/pam_modules.h>
#include <security/_pam_macros.h>
#include <security/pam_ext.h>

#ifdef HAVE_PAM_CHECK_USER_IN_PASSWD
#include <security/pam_modutil.h>
#endif

/* argument parsing */
#define PAM_DEBUG_ARG       0x0001

struct module_options {
        pwquality_settings_t *pwq;
};

#define CO_RETRY_TIMES  1

#define PATH_PASSWD "/etc/passwd"

static int
_pam_parse (pam_handle_t *pamh, struct module_options *opt,
            int argc, const char **argv)
{
        int ctrl = 0;
        int rv;
        pwquality_settings_t *pwq;
        void *auxerror;
        char buf[PWQ_MAX_ERROR_MESSAGE_LEN];

        pwq = pwquality_default_settings();
        if (pwq == NULL)
                return -1;

        /* just log error here */
        if ((rv=pwquality_read_config(pwq, NULL, &auxerror)) != 0)
                pam_syslog(pamh, LOG_ERR,
                        "Reading pwquality configuration file failed: %s",
                        pwquality_strerror(buf, sizeof(buf), rv, auxerror));

        /* step through arguments */
        for (ctrl = 0; argc-- > 0; ++argv) {
                if (!strcmp(*argv, "debug"))
                        ctrl |= PAM_DEBUG_ARG;
                else if (!strncmp(*argv, "type=", 5))
                        pam_set_item (pamh, PAM_AUTHTOK_TYPE, *argv+5);
                else if (!strncmp(*argv, "difignore=", 10)) {
                        /* ignored for compatibility with pam_cracklib */
                } else if (!strncmp(*argv, "reject_username", 15)) {
                        /* ignored for compatibility with pam_cracklib */
                } else if (!strncmp(*argv, "authtok_type", 12)) {
                        /* for pam_get_authtok, ignore */;
                } else if (!strncmp(*argv, "use_authtok", 11)) {
                        /* for pam_get_authtok, ignore */;
                } else if (!strncmp(*argv, "use_first_pass", 14)) {
                        /* for pam_get_authtok, ignore */;
                } else if (!strncmp(*argv, "try_first_pass", 14)) {
                        /* for pam_get_authtok, ignore */;
                } else if (pwquality_set_option(pwq, *argv)) {
                        pam_syslog(pamh, LOG_ERR,
                                "pam_parse: unknown or broken option; %s", *argv);
                }
         }

         opt->pwq = pwq;

         return ctrl;
}

static int
check_local_user (pam_handle_t *pamh,
                  const char *user)
{
#ifdef HAVE_PAM_CHECK_USER_IN_PASSWD
        return pam_modutil_check_user_in_passwd(pamh, user, NULL) == PAM_SUCCESS;
#else
        struct passwd pw, *pwp;
        char buf[4096];
        int found = 0;
        FILE *fp;
        int errn;

        fp = fopen(PATH_PASSWD, "r");
        if (fp == NULL) {
                pam_syslog(pamh, LOG_ERR, "unable to open %s: %s",
                           PATH_PASSWD, pam_strerror(pamh, errno));
                return -1;
        }

        for (;;) {
                errn = fgetpwent_r(fp, &pw, buf, sizeof (buf), &pwp);
                if (errn == ERANGE) {
                        pam_syslog(pamh, LOG_WARNING, "%s contains very long lines; corrupted?",
                                   PATH_PASSWD);
                        /* we can continue here as next call will read further */
                        continue;
                }
                if (errn != 0)
                        break;
                if (strcmp(pwp->pw_name, user) == 0) {
                        found = 1;
                        break;
                }
        }

        fclose (fp);

        if (errn != 0 && errn != ENOENT) {
                pam_syslog(pamh, LOG_ERR, "unable to enumerate local accounts: %s",
                           pam_strerror(pamh, errn));
                return -1;
        } else {
                return found;
        }
#endif
}

PAM_EXTERN int
pam_sm_chauthtok(pam_handle_t *pamh, int flags,
                 int argc, const char **argv)
{
        int ctrl;
        struct module_options options;
        int retry_times = CO_RETRY_TIMES;

        memset(&options, 0, sizeof(options));

        ctrl = _pam_parse(pamh, &options, argc, argv);
        if (ctrl < 0)
                return PAM_BUF_ERR;

        pwquality_get_int_value(options.pwq, PWQ_SETTING_RETRY_TIMES, &retry_times);
        if (retry_times < 1)
                retry_times = CO_RETRY_TIMES;


        if (flags & PAM_PRELIM_CHECK) {
                /* Check for passwd dictionary
                 * We cannot do that, since the original path is compiled
                 * into the cracklib library and we don't know it.
                 */
                pwquality_free_settings(options.pwq);
                return PAM_SUCCESS;
        } else if (flags & PAM_UPDATE_AUTHTOK) {
                int retval;
                const void *oldtoken;
                const char *user;
                int tries;

                retval = pam_get_user(pamh, &user, NULL);
                if (retval != PAM_SUCCESS || user == NULL) {
                        if (ctrl & PAM_DEBUG_ARG)
                                pam_syslog(pamh, LOG_ERR, "Can not get username");
                        pwquality_free_settings(options.pwq);
                        return PAM_AUTHTOK_ERR;
                }

                retval = pam_get_item(pamh, PAM_OLDAUTHTOK, &oldtoken);
                if (retval != PAM_SUCCESS) {
                        if (ctrl & PAM_DEBUG_ARG)
                                pam_syslog(pamh, LOG_ERR, "Can not get old passwd");
                        oldtoken = NULL;
                }

                tries = 0;
                while (tries < retry_times) {
                        void *auxerror;
                        const char *newtoken = NULL;
                        int local_users_only = 0;

                        tries++;

                        /* Planned modus operandi:
                         * Get a passwd.
                         * Verify it against libpwquality.
                         * If okay get it a second time.
                         * Check to be the same with the first one.
                         * set PAM_AUTHTOK and return
                         */

                        retval = pam_get_authtok_noverify(pamh, &newtoken, NULL);
                        if (retval != PAM_SUCCESS || newtoken == NULL) {
                                if (retval == PAM_AUTHTOK_ERR || newtoken == NULL)
                                        pam_syslog(pamh, LOG_INFO, "user aborted password change");
                                else
                                        pam_syslog(pamh, LOG_ERR, "pam_get_authtok_noverify returned error: %s",
                                               pam_strerror(pamh, retval));
                                pwquality_free_settings(options.pwq);
                                return PAM_AUTHTOK_ERR;
                        }
                        pwquality_get_int_value(options.pwq, PWQ_SETTING_LOCAL_USERS, &local_users_only);
                        if ((ctrl & PAM_DEBUG_ARG) && local_users_only) {
                                pam_syslog(pamh, LOG_INFO, "Applying password quality checks to local users only");
                        }

                        if (local_users_only && check_local_user (pamh, user) == 0) {
                                /* skip the check if a non-local user */
                                retval = 0;
                        } else {
                                /* now test this passwd against libpwquality */
                                retval = pwquality_check(options.pwq, newtoken, oldtoken, user, &auxerror);
                        }

                        if (retval < 0) {
                                const char *msg;
                                char buf[PWQ_MAX_ERROR_MESSAGE_LEN];
                                int enforcing = 1;
                                int enforce_for_root = 0;

                                msg = pwquality_strerror(buf, sizeof(buf), retval, auxerror);
                                if (ctrl & PAM_DEBUG_ARG)
                                        pam_syslog(pamh, LOG_DEBUG, "bad password: %s", msg);
                                pwquality_get_int_value(options.pwq, PWQ_SETTING_ENFORCING, &enforcing);
                                pwquality_get_int_value(options.pwq, PWQ_SETTING_ENFORCE_ROOT, &enforce_for_root);
                                if (enforcing && (getuid() || enforce_for_root ||
                                    (flags & PAM_CHANGE_EXPIRED_AUTHTOK))) {
                                        pam_error(pamh, _("BAD PASSWORD: %s"), msg);
                                        pam_set_item(pamh, PAM_AUTHTOK, NULL);
                                        continue;
                                } else {
                                        pam_info(pamh, _("WARNING: Weak password: %s\nPassword quality requirements are not enforced."), msg);
                                }
                        } else {
                                if (ctrl & PAM_DEBUG_ARG)
                                        pam_syslog(pamh, LOG_DEBUG, "password score: %d", retval);
                        }

                        retval = pam_get_authtok_verify(pamh, &newtoken, NULL);
                        if (retval != PAM_SUCCESS || newtoken == NULL) {
                                pam_set_item(pamh, PAM_AUTHTOK, NULL);
                                if (retval == PAM_TRY_AGAIN)
                                        continue;
                                if (retval == PAM_AUTHTOK_ERR || newtoken == NULL)
                                        pam_syslog(pamh, LOG_INFO, "user aborted password change");
                                else
                                        pam_syslog(pamh, LOG_ERR, "pam_get_authtok_verify returned error: %s",
                                               pam_strerror(pamh, retval));
                                pwquality_free_settings(options.pwq);
                                return PAM_AUTHTOK_ERR;
                        }

                        pwquality_free_settings(options.pwq);
                        return PAM_SUCCESS;
                }

                pwquality_free_settings(options.pwq);
                pam_set_item (pamh, PAM_AUTHTOK, NULL);

                /* if we have only one try, we can use the real reason,
                 * else say that there were too many tries. */
                if (retry_times > 1)
                        return PAM_MAXTRIES;
                else
                        return PAM_AUTHTOK_ERR;
        } else {
                pwquality_free_settings(options.pwq);
                if (ctrl & PAM_DEBUG_ARG)
                        pam_syslog(pamh, LOG_NOTICE, "UNKNOWN flags setting %02X",flags);
        }

        return PAM_SERVICE_ERR;
}



#ifdef PAM_STATIC
/* static module data */
struct pam_module _pam_pwquality_modstruct = {
     "pam_pwquality",
     NULL,
     NULL,
     NULL,
     NULL,
     NULL,
     pam_sm_chauthtok
};
#endif

/*
 * Copyright (c) Cristian Gafton <gafton@redhat.com>, 1996.
 *                                              All rights reserved
 * Copyright (c) Red Hat, Inc, 2011, 2012
 * Copyright (c) Tomas Mraz <tm@t8m.info>, 2011
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, and the entire permission notice in its entirety,
 *    including the disclaimer of warranties.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior
 *    written permission.
 *
 * ALTERNATIVELY, this product may be distributed under the terms of
 * the GNU General Public License version 2 or later, in which case the
 * provisions of the GPL are required INSTEAD OF the above restrictions.
 *
 * THIS SOFTWARE IS PROVIDED `AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The following copyright was appended for the long password support
 * added with the libpam 0.58 release:
 *
 * Modificaton Copyright (c) Philip W. Dalrymple III <pwd@mdtsoft.com>
 *       1997. All rights reserved
 *
 * THE MODIFICATION THAT PROVIDES SUPPORT FOR LONG PASSWORD TYPE CHECKING TO
 * THIS SOFTWARE IS PROVIDED `AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */
